spec/formatting.md

```diff
@@ -246,65 +247,81 @@ the following steps are taken:
 
 ### Fallback Resolution
 
-A **_fallback value_** is the resolved value emitted when an _expression_ cannot be resolved.
+A **_fallback value_** is the resolved value for an _expression_ that fails to resolve.
 
 An _expression_ fails to resolve when:
 
-- A _variable_ _operand_ fails to resolve.
+- A _variable_ used as an _operand_ (with or without an _annotation_) fails to resolve.
+  * Note that this does not include a _variable_ used as an _option_ value.
 - A _function_ _annotation_ fails to resolve.
 - A _private-use_ _annotation_ is unsupported by the implementation or if
   a _private-use_ _annotation_ fails to resolve.
 - The _expression_ has a _reserved_ _annotation_.
 
 The _fallback value_ depends on the contents of the _expression_:
 
-- _expression_ with _literal_ _operand_: U+007C VERTICAL LINE `|`
-  followed by the value of the Literal,
-  and then by U+007C VERTICAL LINE `|`.
-  The same representation is used for both _quoted_ and _unquoted_ values.
+- _expression_ with _annotation_ and no _operand_:
+  - If the _annotation_ is a _function_ _annotation_:
+    the _annotation_ starting sigil followed by its _identifier_
 
-  > Examples: `|your horse|`, `|42|`
+    > Examples:
+    > In a context where `:func` fails to resolve, `{:func}` resolves to the _fallback value_ `:func`.
+    > In a context where `:ns:func` fails to resolve, `{:ns:func}` resolves to the _fallback value_ `:ns:func`.
 
-- _expression_ with _variable_ _operand_: U+0024 DOLLAR SIGN `$`
-  followed by the _variable_ _name_ of the _operand_
+  - If the non-_function_ _annotation_ consists of a _sigil_-prefixed _identifier_
```

Comment:
A "non-_function_ _annotation_" is the same as `reserved-annotation`+`private-use-annotation`. Is there some reason to be elliptical about it?

```suggestion
If the _annotatation_ is a _reserved annotation_ or a _private-use annotation_
and the `reserved` or `private-use` sigil is immediately followed by a character
sequence that is an _identifier_, 
make the _fallback value_ the sigil followed by that _identifier_
```

Reply:
I figure it's pretty much "function" vs. "other", because there's no differentiation here between private-use vs. reserved annotations. But your suggestion does more than just rephrase, because the instances of the latter do not necessarily start with a sigil-prefixed _identifier_ (e.g., consider `{@ reserved}` and `{>}` and `{&|quoted|}`).

Reply:
Those instances would not match my text and they would emit the logo. What's our intention here? Do we want to pass reserved/private garbage unmolested? Or turn them into the logo? My reading of the current text is that we make them into the logo unless they look like `{^identifier something something}` where we emit `{^identifier}` for some reason.

Reply:
[Current text](https://github.com/unicode-org/message-format-wg/blob/main/spec/formatting.md#fallback-resolution) is incoherent—the predicate of "_expression_ with no _operand_: the _function_ starting sigil followed by its _identifier_" covers non-function expressions, but the consequent cannot be applied to them. That can be fixed by always having such expressions fall back to the logo, or (as I have done here) by applying function-like treatment to those that use function-like syntax. But I consider it important for "function-like" to be defined narrowly enough to exclude e.g. `{@spaceless|quoted|}`.

Reply:
It's almost coherent except for the possibility of reserved/private... which is what we're fixing here. Because `reserved-annotation` and `private-use-annotation` are permissive about content and because we don't have a construct for parsing _inside_ them in the ABNF (nor, I think, should we...) I think emitting the logo makes some sense.

Putting myself in the shoes of a user, this is the rendered output of a message containing some placeholders that my implementation doesn't understand. If we emit their contents, yes, that might assist in debugging. But if shown to users it might contain code internal stuff. The logo basically is an admission that your implementation has no idea what this placeholder means. Does that make sense?

Note: it's probably important that we emit _something_ as a way of preventing sneak attacks such as `{{<scr{@omit me}ipt href=...>}}`

Reply:
There's a bit further down in the spec that probably ought to be highlighted here as well: https://github.com/unicode-org/message-format-wg/blob/3f35bee287671f5162d8599fd0ee85bbf7682759/spec/formatting.md?plain=1#L591-L595

The intent with that is that even though our default fallback must guard against e.g. `{$user :name}` having a failure mode where a custom stringifier of the `$user` value gets called, dumping personal data into the output, there are still plenty of places where a better fallback than the default can and should be defined.

But it does mean that the space of fallback values cannot be considered exhaustive, and so a reserved expression producing `{�}` fallback now should not be considered a restriction on what might happen if we introduce a new meaning for its sigil later. Also, we should add language somewhere allowing an implementation to define separately the fallback behaviour for private-use sigils.

Put together, I think we _could_ default to `{�}` fallback for reserved and private-use, but make it clear(er) that later spec versions or implementations supporting private-use can and should define something better.

Reply:
I also think that it is important to indicate to the user that a fallback happened : that is, what is displayed to the user is **_not_** the intended result. One possibility is prefixing and/or suffixing with �.

Reply:
@macchiati Perhaps? Presumably the presence of curly-bracketed garbage is an indicator. But having &#xfffd; present consistently would make it easier to find and easier to check for. If we did this, I'd make it a prefix so that we'd have fallbacks like:

```
{�}
{�$var}
{�|42.7 some literal|}
{�:function}
{�^unsupported}
```